# Análisis de viabilidad del TFM
En `df_fases_MES`tengo la **fase**, **máquina dominante**, y el **grupo/familia de esa máquina**. Con eso puedo montar un  cruce fase × máquina dentro de cada familia, ahi es donde vive la decisión de asignación.

Entonces, la elección real no es "esta fase fue a esta máquina" (eso siempre pasa). La elección existe cuando fases del mismo tipo pudieron ir a máquinas distintas de la misma familia.   
Por eso el análisis interesante es dentro de cada grupo: `¿las fases de una familia se reparten entre sus varias máquinas, o cada tipo de fase va siempre a la misma?`

In [2]:
import os
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import pyarrow as pa
import pickle

## Revisamos lo visto hasta ahora:


In [3]:
# importo variables locales
with open("res/procesed/variables/1.Tratamiento_Imputaciones.pkl", "rb") as f:
    variables = pickle.load(f)
df_inicial = variables["Imputaciones iniciales"]
df_intervalo_timepos = variables["Imputaciones dentro del intervalo temporal"]

with open("res/procesed/variables/2.Tratamiento_Machines.pkl", "rb") as f:
    variables = pickle.load(f)
df_maquinas_inicio = variables["Máquinas iniciales"]
df_machines_MES = variables["Maquinas activas en MES"]
df_maquinas_productivas = variables["Máquinas productivas"]
df_maquinas_posibles = variables["Máquinas con suficientes imputaciones"]
df_maquina_monomaquina = variables["Máquinas en familias validas"]
df_maquinas_grupos = variables["Familias iniciales"]
resumen_familia = variables["Familias finales"]

with open("res/procesed/variables/3.Tratamiento_Imputaciones.pkl", "rb") as f:
    variables = pickle.load(f) 
df_imputations_productivas = variables["Imputaciones en máquinas productivas"]
total_fases_productivas = variables["Fases en máquinas productivas"]
imputaciones_maquina_dominante = variables["Fases válidas (dominante >80%)"]
df_fases_sin_mal_cerrar = variables["Fases sin imputaciones mal cerradas"]

with open("res/procesed/variables/4.Tratamiento_Datos_OF.pkl", "rb") as f:
    variables = pickle.load(f)
df_OF_iniciales = variables["OF iniciales"]
df_OF_intervalo_tiempos = variables["OF en intervalo temporal"]


with open("res/procesed/variables/5.Tratamiento_Datos_FASES.pkl", "rb") as f:
    variables = pickle.load(f)
df_FASE_iniciales = variables["Fases iniciales"]
df_FASE_sin_externas = variables["Fases sin externas"]
df_fase_en_OF_validad = variables["Fases en OF válidas"]
df_FASE_merge_imputaciones = variables["Fases con imputaciones validadas"]
df_fases_tiempos_teoricos = variables["Fases con tiempos teóricos"]
df_fases_trabajos_validos = variables["Fases trabajos válidos"]
OF_finales = variables["OF finales"]
        

In [4]:
# Tablas de máquinas y familias
tabla_resultados = pd.DataFrame({
    "Máquinas iniciales": [df_maquinas_inicio],
    "Maquinas activas en MES": [df_machines_MES],
    "Máquinas productivas": [df_maquinas_productivas],
    "Máquinas con suficientes imputaciones": [df_maquinas_posibles],
    "Máquinas en familias validas": [df_maquina_monomaquina],
    "Familias iniciales": [df_maquinas_grupos],
    "Familias finales": [resumen_familia]
}).T

tabla_resultados = tabla_resultados.rename(columns={0: "Número de instancias"})

print("=======================================================")
print("Evolución tablas de las máquinas y familias:")
print("=======================================================")
display(tabla_resultados)
print("=======================================================")

# Tablas de imputaciones
tabla_resultados = pd.DataFrame({
    "Imputaciones iniciales": [df_inicial],
    "Imputaciones dentro del intervalo temporal": [df_intervalo_timepos],
    "Imputaciones en máquinas productivas": [df_imputations_productivas], 
}).T

tabla_resultados = tabla_resultados.rename(columns={0: "Número de imputaciones"})

print("=======================================================")
print("Evolución de las imputaciones:")
print("=======================================================")
display(tabla_resultados)
print("=======================================================")

# Tablas de fases
tabla_resultados = pd.DataFrame({
    "Fases iniciales": [df_FASE_iniciales], 
    "Fases sin externas": [df_FASE_sin_externas], 
    "Fases en OF válidas": [df_fase_en_OF_validad], 
       "Fases en máquinas productivas": [total_fases_productivas], 
        "Fases válidas (dominante >80%)": [imputaciones_maquina_dominante],
        "Fases sin imputaciones mal cerradas": [df_fases_sin_mal_cerrar], 
    "Fases con imputaciones validadas": [df_FASE_merge_imputaciones], 
    "Fases con tiempos teóricos": [df_fases_tiempos_teoricos], 
    "Fases de trabajos válidos": [df_fases_trabajos_validos]     
}).T

tabla_resultados = tabla_resultados.rename(columns={0: "Número de fases"})

print("=======================================================")
print("Evolución de las fases:")
print("=======================================================")
display(tabla_resultados)
print("=======================================================")

# Tabla de OF
tabla_resultados = pd.DataFrame({
    "OF iniciales": [df_OF_iniciales], 
    "OF en intervalo temporal": [df_OF_intervalo_tiempos], 
    "OF finales": [OF_finales] 
}).T

tabla_resultados = tabla_resultados.rename(columns={0: "Número de OF"})

print("=======================================================")
print("Evolución de las OF:")
print("=======================================================")
display(tabla_resultados)
print("=======================================================")

Evolución tablas de las máquinas y familias:


,Número de instancias
Máquinas iniciales,120
Maquinas activas en MES,119
Máquinas productivas,76
Máquinas con suficientes imputaciones,38
Máquinas en familias validas,32
Familias iniciales,57
Familias finales,9


Evolución de las imputaciones:


,Número de imputaciones
Imputaciones iniciales,196576
Imputaciones dentro del intervalo temporal,140485
Imputaciones en máquinas productivas,59570


Evolución de las fases:


,Número de fases
Fases iniciales,116055
Fases sin externas,107542
Fases en OF válidas,69986
Fases en máquinas productivas,19149
Fases válidas (dominante >80%),18548
Fases sin imputaciones mal cerradas,17389
Fases con imputaciones validadas,16383
Fases con tiempos teóricos,14995
Fases de trabajos válidos,14759


Evolución de las OF:


,Número de OF
OF iniciales,18106
OF en intervalo temporal,11983
OF finales,6551


In [5]:
df_final = pd.read_parquet("res//procesed/df_final.parquet")
pd.set_option("display.max_columns", None)
pd.set_option("display.width", None)   
display(df_final.head(1))

,SmFaseUid,MaquinaDominante,fecha_fase,smOrdenId,smMaterialId,smFaseNme,smOrdenNme,MinutosReal_efectivo,MinutosPreparacion_efectivo,MinutosProduccion_efectivo,MinutosReal_total,MinutosPreparacion_total,MinutosProduccion_total,dedicacion_media,n_imputaciones,IDMaquina,NombreMaquina,IDGrupo,CodigoGrupo,NombreGrupo,IDMOTask,IDManufacturingOrder,Description,PreparationTime,ExecutionTime,TimeUnit,ProductEntry,MaterialExit,IDUsualTask,Quantity,ManualStartDate,RealEndDate,MESEntityId,Código Artículo,Descripción Artículo,CodProductLine,Nombre Línea,CodProductFamily,Nombre Família
0,00013c31-3895-4370-bb3e-4255371e8d8f,TN09,2025-10-02 03:12:05.010,OF2509545,PE1JNCX0012,TPP,Portacápsulas F704501049704(150713),41.76667,41.76667,0.0,41.76667,41.76667,0.0,50.0,2,0bab3b63-5d31-4de3-85e2-89dc5a3e9bd5,TPP DOOSAN PUMA 240 (TPP) Torn C.N.C,27451f15-090c-42a6-82d6-98b7451db3db,TPU,Torns Puma (Taller COURES),ee76d685-a4cc-47d4-b418-6e33d2ae2921,26a3fecc-120e-49bf-9e77-720d35355e3d,Mecanitzat al Torn Puma Petit,6.0,15.0,1.0,0.0,0.0,29783995-2f56-48be-9664-461e6f0b402a,3.0,2025-10-13 12:55:00.000,2025-10-02 00:00:00.000,00013c31-3895-4370-bb3e-4255371e8d8f,PE1JNCX0012,Elektrodenaufnamhe F704501049704(DS086228),P,Portalectrodes,PE,Portalectrodes


Cuántas máquinas y cuántas fases tiene cada familia:

In [6]:
resumen_familia = df_final.groupby("NombreGrupo").agg(
    n_fases=("SmFaseUid", "nunique"),
    n_maquinas=("MaquinaDominante", "nunique"),
).sort_values("n_fases", ascending=False)

display(resumen_familia)
print(f"Total grupos de máquinas: {len(resumen_familia)}")

,n_fases,n_maquinas
NombreGrupo,,
Torns Puma (Taller COURES),4088,2
Màquines de Taladrat Profund,3604,2
Centres de Mecanitzat Vertical COURES,2636,4
Torns Control Numèric,1641,4
Centres de Mecanitzat Horitzontal COURES,1242,3
Centre de Mecanitzat 5 Eixos petites,553,4
Centres de Mecanitzat Vertical ALUMINIS,507,3
Premses Conformats,314,2
Màquines Secció Flexibles,174,4


Total grupos de máquinas: 9


Toca observar si las imputaciones se reparten entre las máquinas de la família.

In [7]:
familias_analizables = set(df_final["NombreGrupo"])

for familia in familias_analizables:
    sub = df_final[df_final["NombreGrupo"] == familia]
    reparto = sub["MaquinaDominante"].value_counts(normalize=True)
    print(f"{familia[:45]:45s} | fases: {len(sub):5d} | máq. principal: {reparto.iloc[0]:.0%} | nº máq. usadas: {(reparto > 0.001).sum()}")

Centre de Mecanitzat 5 Eixos petites          | fases:   553 | máq. principal: 80% | nº máq. usadas: 4
Centres de Mecanitzat Vertical ALUMINIS       | fases:   507 | máq. principal: 65% | nº máq. usadas: 3
Centres de Mecanitzat Vertical COURES         | fases:  2636 | máq. principal: 49% | nº máq. usadas: 4
Torns Puma (Taller COURES)                    | fases:  4088 | máq. principal: 79% | nº máq. usadas: 2
Màquines Secció Flexibles                     | fases:   174 | máq. principal: 98% | nº máq. usadas: 4
Centres de Mecanitzat Horitzontal COURES      | fases:  1242 | máq. principal: 73% | nº máq. usadas: 3
Premses Conformats                            | fases:   314 | máq. principal: 56% | nº máq. usadas: 2
Màquines de Taladrat Profund                  | fases:  3604 | máq. principal: 89% | nº máq. usadas: 2
Torns Control Numèric                         | fases:  1641 | máq. principal: 34% | nº máq. usadas: 4


In [8]:
familias_analizables = set(df_final["NombreGrupo"])

resultado = []
for familia in familias_analizables:
    subset = df_final[df_final["NombreGrupo"] == familia]
    n_fases_distintas = subset["Description"].nunique()
    resultado.append({"Nombre grupo": familia, "Cantidad tipos fases": n_fases_distintas})

contando_fases = pd.DataFrame(resultado)
display(contando_fases)

,Nombre grupo,Cantidad tipos fases
0,Centre de Mecanitzat 5 Eixos petites,36
1,Centres de Mecanitzat Vertical ALUMINIS,10
2,Centres de Mecanitzat Vertical COURES,42
3,Torns Puma (Taller COURES),26
4,Màquines Secció Flexibles,2
5,Centres de Mecanitzat Horitzontal COURES,33
6,Premses Conformats,3
7,Màquines de Taladrat Profund,6
8,Torns Control Numèric,65


Vamos a trabajar con tiempo total, en las imputaciones aveces todo el peso se queda en preparación o producción, sin separlo, esto se debe a una mala disciplina entre los operarios. 

In [9]:
# ¿cuántas fases tienen teórico en una parte pero real 0 en esa misma parte?
prep_teorico_sin_real = ((df_final["PreparationTime"] > 0) & (df_final["MinutosPreparacion_efectivo"] == 0)).sum()
prod_teorico_sin_real = ((df_final["ExecutionTime"] > 0) & (df_final["MinutosProduccion_efectivo"] == 0)).sum()

print(f"Fases con teórico de preparación pero real de preparación = 0: {prep_teorico_sin_real}")
print(f"Fases con teórico de producción pero real de producción = 0: {prod_teorico_sin_real}")
print(f"Total fases: {len(df_final)}\n")

Fases con teórico de preparación pero real de preparación = 0: 369
Fases con teórico de producción pero real de producción = 0: 12089
Total fases: 14759



Casi todas las fases tienen teórico en una parte pero real 0 en esa misma parte, significa que la separación preparación/producción en el MES no es fiable: el trabajo se imputa a una casilla u otra sin correspondencia con cómo lo definió el ERP.

In [10]:
# Inspeccionar las maquinas de cada familia:
maquinas_por_familia = df_final.groupby(["NombreGrupo", "MaquinaDominante"]).size()

for familia, maquinas in maquinas_por_familia.groupby(level=0):
    print(f"{familia}:")
    print(maquinas.droplevel(0).sort_values(ascending=False).to_string())
    print()

Centre de Mecanitzat 5 Eixos petites:
MaquinaDominante
CM12    443
CM15     86
CM17     23
CM18      1

Centres de Mecanitzat Horitzontal COURES:
MaquinaDominante
CM05    910
CM01    212
CM07    120

Centres de Mecanitzat Vertical ALUMINIS:
MaquinaDominante
CM08    332
CM10    157
CM04     18

Centres de Mecanitzat Vertical COURES:
MaquinaDominante
CM06    1287
CM14     708
CM16     602
CM19      39

Màquines Secció Flexibles:
MaquinaDominante
FL01    171
CM11      1
CM13      1
MA02      1

Màquines de Taladrat Profund:
MaquinaDominante
TP02    3190
TP01     414

Premses Conformats:
MaquinaDominante
PH04    175
PH03    139

Torns Control Numèric:
MaquinaDominante
TN08    556
TN10    519
TN05    433
TN07    133

Torns Puma (Taller COURES):
MaquinaDominante
TN09    3224
TN04     864



## Decisión de Viabilidad 
Tras el proceso de limpieza y depuración realizado, se dispone de un bloque sólido de familias: 9 familias con elección real y volumen suficiente.

Este resultado cierra el análisis de viabilidad, con una conclusión positiva.

- Torns Control Numèric — 34% de concentración en la máquina principal, 4 máquinas utilizadas. El trabajo se reparte de forma equilibrada entre las cuatro, lo que valida la fusión de grupos realizada y constituye el ejemplo más claro de elección real. 1.641 fases.
- Centres Vertical COURES — 49% de concentración, 4 máquinas utilizadas. Reparto saludable sobre 2.636 fases.
- Premses Conformats — 56% de concentración, 2 máquinas utilizadas. Reparto equilibrado entre las dos máquinas activas, con un volumen moderado (314 fases).
- Centres Vertical ALUMINIS — 65% de concentración, 3 máquinas utilizadas. Reparto adecuado sobre 507 fases.
- Centres Horitzontal COURES — 73% de concentración, 3 máquinas utilizadas. Reparto aceptable sobre 1.242 fases.

Estas cinco familias constituyen el núcleo sólido del análisis: presentan una decisión de asignación efectiva y un reparto real del trabajo entre máquinas.

- Torns Puma — 79% de concentración, 2 máquinas utilizadas. Aunque la máquina principal concentra la mayor parte del trabajo, el 21% restante sobre 4.088 fases representa un volumen relevante en la segunda máquina.
- Centre 5 Eixos petites — 80% de concentración, 4 máquinas utilizadas. La máquina principal domina, pero existe reparto entre las cuatro máquinas. 553 fases.
- Màquines de Taladrat Profund — 89% de concentración, 2 máquinas utilizadas. Una máquina concentra casi todo el trabajo, aunque el 11% restante sobre 3.604 fases sigue representando un volumen relevante en la segunda.

Estas tres familias se consideran frontera: se incluyen en el análisis, si bien podrían quedar excluidas según evolucione el estudio.

- Màquines Secció Flexibles — 98% de concentración, y solo 174 fases. La máquina principal concentra casi la totalidad del trabajo sobre un volumen bajo, por lo que la elección resulta prácticamente inexistente. Se propone excluir esta familia del análisis de idoneidad, o mencionarla explícitamente como familia sin elección efectiva.

## Conclusión
De las familias productivas con múltiples máquinas, la mayoría presentan un reparto efectivo del trabajo entre sus máquinas, lo que confirma que la asignación fase-máquina constituye una decisión real y no determinista en el taller.

NOTA PERSONAL:
puede que haya criterios legítimos (tamaño de pieza, disponibilidad, precisión requerida) que hacen que cierta máquina sea la natural para la mayoría de fases, reservando las otras para casos concretos. Eso también es idoneidad, y de hecho es lo interesante que tu modelo podría capturar. Así que no descartes las de principal alta a la ligera; el matiz es "poca elección" en Secció Flexibles (por volumen + concentración), no en las demás.

## Justificación
Análisis de viabilidad: existencia de una decisión de asignación real

Antes de abordar el modelado, resulta imprescindible verificar una premisa sobre la que se sostiene todo el trabajo: que la asignación de una fase a una máquina constituye efectivamente una decisión con alternativas, y no un proceso determinista en el que cada fase se ejecuta necesariamente en un único recurso posible. Si no existiera tal elección, no habría idoneidad que evaluar ni asignación que optimizar, y el problema carecería de sentido.

Para comprobarlo, se ha analizado el conjunto de familias de máquina productivas resultantes del proceso de depuración, cada una de las cuales agrupa máquinas funcionalmente equivalentes entre las que el taller puede repartir el trabajo. Se ha exigido, como condición necesaria, que una familia disponga de al menos dos máquinas; las familias constituidas por una única máquina se excluyeron de este análisis por no ofrecer alternativa alguna de asignación. Sobre las familias restantes, se ha examinado no solo cuántas máquinas podrían ejecutar sus fases, sino en qué medida el trabajo se reparte realmente entre ellas, calculando la proporción de fases que concentra la máquina más utilizada de cada familia.

Los resultados confirman la existencia de una decisión de asignación efectiva. En la mayoría de las familias analizadas, la máquina principal concentra entre el 34 % y el 73 % de las fases, evidenciando que el trabajo se distribuye de forma genuina entre varios recursos y que, por tanto, existe una elección real en la asignación. Casos como el de los tornos de control numérico (34 % en la máquina principal, con las fases repartidas entre cuatro máquinas) o los centros de mecanizado vertical de cobre (49 %) constituyen la evidencia más clara de que la asignación no es determinista. Se observaron asimismo familias con una concentración elevada en una única máquina (superior al 85 %), que, si bien disponen de alternativas, presentan una elección efectiva más limitada; estas se tratan con cautela en el análisis posterior.

En conjunto, este análisis valida empíricamente la premisa central del trabajo: existe un problema de asignación fase-máquina con capacidad de elección real, sobre un volumen de datos suficiente, que un modelo de idoneidad puede aprender a caracterizar y optimizar.

En conjunto, tras el proceso completo de depuración, el análisis se sostiene sobre un total de 14.759 fases distribuidas en las familias de máquina productivas con capacidad de elección, un volumen que proporciona una base de datos amplia y suficiente para el modelado posterior. Este análisis valida empíricamente la premisa central del trabajo: existe un problema de asignación fase-máquina con capacidad de elección real, sobre un volumen de datos suficiente, que un modelo de idoneidad puede aprender a caracterizar y optimizar.

Con esto la sección de viabilidad queda completa y con las cifras reales: a partir de 140.485 imputaciones en crudo, el proceso de depuración ha resultado en una población de 14.759 fases analizables, cada una con su máquina, su familia y sus tiempos reales.